plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs

```diff
@@ -0,0 +1,501 @@
+    didn't get a minimal counterexample (because shrinking requries well-typed terms)
+    you need to use a different trick. One trick that often works is to add the well-typedness
+    check in the generators - e.g. in `genTerm` you can do something like this:
+    ```
+    genTerm ... = myCheck $ do
+      ...
+      where
+        myCheck gen = do
+          (trm, type) <- gen
+          if notConformingToExpectedInvariant trm type then
+            error (show trm ++ " " ++ show type)
+          else
+            return (trm, type)
+    ```
+-}
+
+-- * Dealing with fresh names
+
+-- | Get the free variables in a type that appear in negative position
+negativeVars :: Type TyName DefaultUni () -> Set TyName
+negativeVars ty = case ty of
+  TyFun _ a b      -> positiveVars a <> negativeVars b
+  TyApp _ a b      -> negativeVars a <> negativeVars b
+  TyLam _ x _ b    -> Set.delete x $ negativeVars b
+  TyForall _ x _ b -> Set.delete x $ negativeVars b
+  TyVar _ _        -> mempty
+  TyBuiltin{}      -> mempty
+  TyIFix{}         -> error "negativeVars: TyIFix"
+
+-- | Get the free variables in a type that appear in positive position
+positiveVars :: Type TyName DefaultUni () -> Set TyName
+positiveVars ty = case ty of
+  TyFun _ a b      -> negativeVars a <> positiveVars b
+  TyApp _ a b      -> positiveVars a <> positiveVars b
+  TyLam _ x _ b    -> Set.delete x $ positiveVars b
+  TyForall _ x _ b -> Set.delete x $ positiveVars b
+  TyVar _ x        -> Set.singleton x
+  TyBuiltin{}      -> mempty
+  TyIFix{}         -> error "positiveVars: TyIFix"
+
+-- | Get the free type variables in a type along with how many
+-- times they occur. The elements of the map are guaranteed to be
+-- non-zero.
+fvTypeBag :: Type TyName DefaultUni () -> Map TyName Int
+fvTypeBag ty = case ty of
+  TyVar _ x        -> Map.singleton x 1
+  TyFun _ a b      -> Map.unionWith (+) (fvTypeBag a) (fvTypeBag b)
+  TyApp _ a b      -> Map.unionWith (+) (fvTypeBag a) (fvTypeBag b)
+  TyLam _ x _ b    -> Map.delete x (fvTypeBag b)
+  TyForall _ x _ b -> Map.delete x (fvTypeBag b)
+  TyBuiltin{}      -> Map.empty
+  TyIFix{}         -> error "fvTypeBag: TyIFix"
+
+-- | Freshen a TyName so that it does not equal any of the names in the set.
+freshenTyName :: Set TyName -> TyName -> TyName
+freshenTyName fvs (TyName (Name x j)) = TyName (Name x i)
+  where i  = succ $ Set.findMax is
+        is = Set.insert j $ Set.insert (toEnum 0) $ Set.mapMonotonic (nameUnique . unTyName) fvs
+
+-- | Get the names and types of the constructors of a datatype.
+constrTypes :: Datatype TyName Name DefaultUni DefaultFun () -> [(Name, Type TyName DefaultUni ())]
+constrTypes (Datatype _ _ xs _ cs) = [ (c, abstr ty) | VarDecl _ c ty <- cs ]
+  where
+    abstr ty = foldr (\ (TyVarDecl _ x k) -> TyForall () x k) ty xs
+
+-- | Get the name and type of the match function for a given datatype.
+matchType :: Datatype TyName Name DefaultUni DefaultFun () -> (Name, Type TyName DefaultUni ())
```

Comment:
I still want this to use the code from the compiler.

Reply:
@michaelpj 

> I still want this to use the code from the compiler.

Two things get in the way:

1. fresh name generation is different in the compiler and the generators. And one fresh name is generated somewhat deeply in the compiler
2. the compiler uses non-`()` annotations and they are somewhat harcdoded

I've addressed both in [this](https://github.com/input-output-hk/plutus/pull/4559/commits/6a74dc3970ea7e5e3804bedc42e7f8d64880c840) commit, please review.

(this request is specifically for @michaelpj)

Reply:
Looks fine!